apisix/plugins/authz-casdoor.lua

```diff
@@ -0,0 +1,164 @@
+            ["Content-Type"] = "application/x-www-form-urlencoded"
+        }
+    })
+
+    if not res then
+        return nil, err
+    end
+    local data, err = core.json.decode(res.body)
+
+    if err or not data then
+        err = "failed to parse casdoor response data: " .. err .. ", body: " .. res.body
+        return nil, nil, err
+    end
+
+    if not data.access_token then
+        return nil, nil,
+               "failed when accessing token: no access_token contained"
+    end
+    if not data.expires_in or data.expires_in == 0 then
+        return nil, nil, "failed when accessing token: invalid access_token"
+    end
+
+    return data.access_token, data.expires_in, nil
+end
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+function _M.access(conf, ctx)
+    local current_uri = ctx.var.uri
+    local session_obj_read, session_present = session.open()
+    -- step 1: check whether hits the callback
+    local m, err = ngx.re.match(conf.callback_url, ".+//[^/]+(/.*)", "jo")
+    if err or not m then
+        core.log.error(err)
+        return 503, err
+    end
+    local real_callback_url = m[1]
+    if current_uri == real_callback_url then
+        if not session_present then
+            err = "no session found"
+            core.log.error(err)
+            return 503, err
+        end
+        local state_in_session = session_obj_read.data.state
+        if not state_in_session then
+            err = "no state found in session"
+            core.log.error(err)
+            return 503, err
+        end
+        local access_token, lifetime, err =
+            fetch_access_token(ctx, conf, state_in_session)
+        if err then
+            core.log.error(err)
+            return 503, err
+        end
+        if access_token then
+            local original_url = session_obj_read.data.original_uri
+            if not original_url then
+                err = "no original_url found in session"
+                core.log.error(err)
+                return 503, err
+            end
+            local session_obj_write = session.new {
+                cookie = {lifetime = lifetime}
+            }
+            session_obj_write:start()
+            session_obj_write.data.access_token = access_token
+            session_obj_write:save()
+            core.response.set_header("Location", original_url)
+            return 302
+        else
+            err = "failed to retrieve access_token"
+            return 503, err
+        end
+    end
+
+    -- step 2: check whether session exists
+    if not (session_present and session_obj_read.data.access_token) then
+        -- session not exists, redirect to login page
+        local state = rand(0x7fffffff)
+        local session_obj_write = session.start()
+        session_obj_write.data.original_uri = current_uri
+        session_obj_write.data.state = state
+        session_obj_write:save()
+        local redirect_url = conf.endpoint_addr ..
```

Comment:
Would be better to use [`ngx.encode_args`](https://github.com/openresty/lua-nginx-module#ngxencode_args) to process this.